packages/cloud_firestore/cloud_firestore_web/lib/document_reference_web.dart

```diff
@@ -3,42 +3,40 @@ part of cloud_firestore_web;
 class DocumentReferenceWeb extends DocumentReference {
   final web.Firestore firestoreWeb;
   final web.DocumentReference delegate;
-  DocumentReferenceWeb(this.firestoreWeb,FirestorePlatform firestore, List<String> pathComponents)
-      : delegate = firestoreWeb.doc(pathComponents.join("/")),super(firestore, pathComponents);
+
+  DocumentReferenceWeb(this.firestoreWeb, FirestorePlatform firestore,
+      List<String> pathComponents)
+      : delegate = firestoreWeb.doc(pathComponents.join("/")),
+        super(firestore, pathComponents);
 
   @override
-  Future<void> setData(Map<String, dynamic> data, {bool merge = false}) {
-    return delegate.set(data, web.SetOptions(merge: merge));
-  }
+  Future<void> setData(Map<String, dynamic> data, {bool merge = false}) =>
+      delegate.set(FieldValueWeb._serverDelegates(data), web.SetOptions(merge: merge));
```

Comment:
I think that for web, this `_serverDelegates` of `data` also needs to take care of the following types:

* Blob
* DocumentReference
* GeoPoint

Otherwise they'll throw when they're passed through the [package:firebase `_jsify`](https://github.com/FirebaseExtended/firebase-dart/blob/master/lib/src/utils.dart#L77-L120), won't they?

(It should be the reverse process of how you build a Blob on the web, from the data returned by the server, which maybe we don't do, and just let the user use the package:firebase data types?)

Reply:
I think so too. I didn't get around to testing `Blob` and `GeoPoint` yet. They will need to the same logic as FieldValue where they hold instances of the platform-specific implementation then we can feed it to the appropriate firebase implementation based on platform.

`DocumentReference` is almost there because there's already platform-specific wrappers for it. We just need to get hold of the instance when dealing the actual firebase library.

Reply:
Nice! Thanks for adding them to your TODO!

I'll try to integrate this into another test app, and see how it goes!